Pipfile

```diff
@@ -4,12 +4,26 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-"aws-cdk.core" = "==1.92.0"
+"aws-cdk.core" = "==1.94.1"
+"aws-cdk.aws-ec2" = "==1.94.1"
+"aws-cdk.aws-ecs" = "==1.94.1"
+"aws-cdk.aws-iam" = "==1.94.1"
+"aws-cdk.aws-ecs-patterns" = "==1.94.1"
+"aws-cdk.aws-secretsmanager" = "==1.94.1"
+"aws-cdk.aws-s3" = "==1.94.1"
+"aws-cdk.aws-ecr" = "==1.94.1"
 
 [dev-packages]
 black = "==20.8b1"
 flake8 = "==3.8.4"
 isort = "==5.7.0"
+prefect = "==0.14.13"
+botocore = "==1.19.52"
+dask-cloudprovider = {version = "==2021.3.1", extras = ["aws"]}
+boto3 = "==1.16.52"
+xarray = "*"
+zarr = "*"
+pangeo-forge = {git = "https://github.com/pangeo-forge/pangeo-forge.git" }
```

Comment:
I'd love to pin this to a Tag, but the last tag was Oct 2020, @rabernat are we aiming on doing more frequent tagging of `pangeo-forge`? Or even doing more frequent PyPi releases? Un-pinned dependencies are giving me the heebeejeebees more and more these days 🤣

Reply:
Yes of course. I think we are ready for a release and then to release very frequently thereafter. Note that the [`meta.yaml` spec](https://github.com/pangeo-forge/roadmap/pull/11/files) includes the pangeo_forge version: each recipe will specify its desired version of the package. Pinning recipes to paneo_forge versions will be especially important as the package evolves.

Reply:
Okay, we'll probably have to make it know _which_ version of pangeo_forge that the bakeries are running as currently we're not dynamically installing our worker instance dependencies so that might be a blocker to some recipes getting onto bakeries.

Could you let us know when a tagged version comes out @rabernat ? Cheers!

Reply:
Is it not possible to dynamically install a specific pangeo_forge version when the recipe is executed?

There is an additional problem that we are currently dependent on an unreleased feature in xarray (https://github.com/pydata/xarray/pull/5065).

Reply:
Currently the worker image is held in ECR independently of the bakery, it is referenced on flow registration, so we could have a repository of available images which cover off released tags of pangeo_forge, but this is at registration. We can overwrite the value at runtime of the flow, but the image still needs to be referenced.

Reply:
"worker image" == "dask worker image"?

If so, we can use the [WorkerPlugin](https://pangeo.io/cloud.html#dask-software-environment) approach to install a package on each worker as it starts up.

Reply:
We'll have to look into whether we can work `WorkerPlugin` into the Prefect setup side.

Though doing this we then run the gauntlet of X time to install Y number of packages on the worker, which will happen for every instance of the worker, which could inevitably add up to a fair chunk of time.

Reply:
> We'll have to look into whether we can work `WorkerPlugin` into the Prefect setup side.

FWIW, I am doing this now from a notebook on on Pangeo cloud and it works: Dask Gateway Cluster + PipWorkerPlugin + Prefect DaskExecutor. In Pangeo Cloud, the dask worker image is just the standard [pangeo-notebook](https://github.com/pangeo-data/pangeo-docker-images/tree/master/pangeo-notebook) image. Installing a few runtime dependencies via pip adds about 2s to worker startup time.

Reply:
My overall point here is that, during development, we should probably expect to have a complex and evolving set of dependencies among certain fast-moving packages (pangeo_forge, xarray, fsspec, etc.), including unreleased versions. The ability to add some extra packages at runtime, on top of a base image with the heavy dependencies (HDF5, etc.), will give us a lot more flexibility. I know that this comes at a cost in terms of possible dependency conflicts. But I feel like this is a price worth paying in order to be at the bleeding edge of the packages we are relying on. Having to build a big docker image every time we make a new release of pangeo_forge (or xarray, or fsspec) could slow things down considerably.

Reply:
@rabernat I concur that the flexibility of using `PipInstall` `WorkerPlugin` trumps any dependency conflicts we'll encounter.  I'm going to merge this PR as is and I'll start a new branch which addresses this and your Conda environments comments below 👍 .  

I might be misunderstanding your approach slightly but I think the main issue on our side will be accessing the `DaskExecutor`'s `client`.  I am guessing that in your case you are using Dask Gateway Cluster to instantiate your cluster and then leverage `get_client` to reference the `client` to register the `WorkerPlugin`, then passing this existing cluster to your Prefect `DaskExecutor`.  In our case we pass a `cluster_class` and some ECS infrastructure information to Prefect's `DaskExecutor` which in turn asynchronously creates our cluster and yields the client https://github.com/PrefectHQ/prefect/blob/master/src/prefect/executors/dask.py#L223-L237 as part of the Flow run.  I don't believe there are currently hooks within the Flow run where we could access the executor's `client` and register the `WorkerPlugin`.   I will try to coordinate with Jim to better understand our options here.  Can you provide a public example of your current usage of Dask Gateway Cluster + PipWorkerPlugin + Prefect DaskExecutor as a reference I can look over?